README.md
![Session SSO Logo depicting a cloud with a fingerprint connected to colored dots representing the multiple SSO providers that are supported by this package](https://assets.jeanlescure.io/session-sso-logo.svg)

# Session SSO
![Tests](https://github.com/simplyhexagonal/session-sso/workflows/tests/badge.svg)

Add this package to your back-end API in order to easily authenticate with Google, Facebook, Github,
or even your custom OAuth2 handled by this same library.

## Open source notice

This project is open to updates by its users, [I](https://github.com/jeanlescure) ensure that PRs are relevant to the community.
In other words, if you find a bug or want a new feature, please help us by becoming one of the
[contributors](#contributors-) ✌️ ! See the [contributing section](#contributing)

## Like this module? ❤

Please consider:

- [Buying me a coffee](https://www.buymeacoffee.com/jeanlescure) ☕
- Supporting Simply Hexagonal on [Open Collective](https://opencollective.com/simplyhexagonal) 🏆
- Starring this repo on [Github](https://github.com/simplyhexagonal/package) 🌟

## New to Single Sign On (SSO) integrations?

We suggest taking a look at the [STRATEGY.md](https://github.com/jeanlescure/session-sso/blob/main/STRATEGY.md) document before moving forward here.

## How to get started using this module?

```ts
import SessionSSO from 'session-sso';

const sso = new SessionSSO({
  // facebook dev credentials
  appId: 'oHPrt6...',
  appSecret: 'O9GGmv3KHJ...',
  // github dev credentials
  clientId: 'wf3s6u...',
  clientSecret: 'AxDmXUPUnH...',
});

// note: google doesn't need dev credentials for SSO verification,
// just on front-end to generate the initial OAuth token

const verifyGoogleResult = await sso.verifySSO({
  provider: 'google',
  authKey: 'ey0PweGlS1FG...', // token returned by `googleUser.getAuthResponse().id_token`
});

const verifyFacebookResult = await sso.verifySSO({
  provider: 'facebook',
  authKey: 'EF25LPJCBAT...', // token returned by `FB.getLoginStatus()` => `response.authResponse.accessToken`
});

const verifyGithubResult = await sso.verifySSO({
  provider: 'github',
  authKey: 'c12fa85efae0236c034b', // auth code placed in url when redirected back from https://github.com/login/oauth/authorize
});

console.log(verifyGoogleResult); // { "email": "user-email-address@gmail.com" }
console.log(verifyFacebookResult); // { "email": "user-email-address@gmail.com" }
console.log(verifyGithubResult); // { "email": "user-email-address@gmail.com" }
```

By default the library will fetch the user's email since it's the most common default scope used by
Google, Facebook, and GitHub's tokens.

If your provider allows you to fetch other scopes with your Client/Dev credentials, then you can
override which user properties are returned by `verifySSO` by passing the a `string[]` using the
`retrieveProperties` option like so:

```ts
await sso.verifySSO({
  provider: 'google',
  authKey: 'eyS1FG0PweGl...',
  retrieveProperties: [
    'email',
    'email_verified',
    'given_name',
    'family_name',
    'locale',
  ]
});

// would return:
// {
//   payload: {
//     "email": "kahless@t-kuv.ma",
//     "email_verified": true,
//     "given_name": "Kahless",
//     "family_name": "-",
//     "locale": "tlh"
//   }
// }
```

You can also set these scopes when instantiating:

```ts
const sso = new SessionSSO({
  // ...
  retrieveProperties: [
    //...
  ],
});
```

**Note:** Doing it this way affects _all_ providers, so make sure all of them have the same naming conventions.

## Create your own Custom SSO Provider

The custom SSO flow was inspired by Google's.

First a user would send their authorization data (i.e. username, passowrd, etc), from here, it's up
to you to generate an authorization promise:

```ts
const authorizationPromise = async () => {
  // your top secret authorization sauce here
  // ...

  // if auth fails
  throw new Error('You cannot pass!');

  // if auth succeeds
  return {
    iss: Math.floor(Date.now() / 1000), // issue date as Epoch number (seconds since 1970)
    exp: Math.floor(Date.now() / 1000) + (60 * 60), // expiration date (1 hour later in this case)
    email: 'user-email-address@gmail.com',
    // ...
  };
};
```

then proceed to produce another promise, this time for the private keys in JSON format:

```ts
const privateKeyPromise: Promise<PEMKeyPromisePayload> = fetch(
  'https://your.static.website/certs.priv.json',
).then((res) => res.json()).then((jpems) => {
  // `res` is a JSON with PEM keys:
  // {
  //   "nnI9yCyGPq3r5zmurEVr05uf": "-----BEGIN RSA PRIVATE KEY-----\nMIIEow...",
  //   "y7I9IXxvGBEOhc9CuBcHIklK": "-----BEGIN RSA PRIVATE KEY-----\nMIIEps..."
  // }
  //
  // For a full example of what this should look like visit: https://www.googleapis.com/oauth2/v1/certs

  // Make it hard for bad actors to reverse-engineer PEM keys by using more than one, randomly
  const randomKeyId = Object.keys(jpems).sort(() => Math.random() - 0.5)[0];

  return {
    kid: randomKeyId,
    pem: jpems[randomKeyId],
  };
});
```

finally send your front-end the resulting authentication token:

```ts
const {
  payload: {
    token, // <== send this
  },
} = await sso.generateSSO({
  authorizationPromise,
  privateKeyPromise,
});
```

now any front-end using your authentication end-point can verify that the token hasn't been forged
by sending it back to your API where you do will do a very similar check as with google, facebook, and github:

```ts
// This is the only extra step. With google, facebook, and github we already know where the verification
// comes from, so we baked it in; but here you get to set your own rule as to where to 
const publicKeyPromise: Promise<PEMKeyPromisePayload> = fetch(
  'https://your.static.website/certs.json',
).then((res) => res.json());

const verifyCustomResult = await sso.verifySSO({
  publicKeyPromise,
  authKey: token as string,
});
```

## Error handling

Both `verifySSO` and `generateSSO` promises will return an object with **only one property**,
either `payload` or `error`.

On success, both `verifySSO` and `generateSSO` will resolve with an object with the `payload`
property, for example:

```js
{
  payload: {
    //...
  }
}
```

On error, both `verifySSO` and `generateSSO` will reject with an object with the `error` property,
for example:

```js
{
  error: "..."
}
```

**Note:** the value of the `error` property is a string.

## 🚨 Where to place this library in your code/API (or "I'm new to SSO, and confused")

We've provided a handy and easy to understand explanation of a proper SSO strategy and where this
library is meant to be used, just take a look at the [STRATEGY.md](https://github.com/jeanlescure/session-sso/blob/main/STRATEGY.md) document.

## Contributing

Yes, thank you! This plugin is community-driven, most of its features are from different authors.
Please update the docs and tests and add your name to the `package.json` file.

## Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://jeanlescure.cr"><img src="https://avatars2.githubusercontent.com/u/3330339?v=4" width="100px;" alt=""/><br /><sub><b>Jean Lescure</b></sub></a><br /><a href="#maintenance-jeanlescure" title="Maintenance">🚧</a> <a href="https://github.com/jeanlescure/session-sso/commits?author=jeanlescure" title="Code">💻</a> <a href="#userTesting-jeanlescure" title="User Testing">📓</a> <a href="https://github.com/jeanlescure/session-sso/commits?author=jeanlescure" title="Tests">⚠️</a> <a href="#example-jeanlescure" title="Examples">💡</a> <a href="https://github.com/jeanlescure/session-sso/commits?author=jeanlescure" title="Documentation">📖</a></td>
    <td align="center"><a href="https://dianalu.design"><img src="https://avatars2.githubusercontent.com/u/1036995?v=4" width="100px;" alt=""/><br /><sub><b>Diana Lescure</b></sub></a><br /><a href="https://github.com/jeanlescure/session-sso/commits?author=DiLescure" title="Documentation">📖</a> <a href="https://github.com/jeanlescure/session-sso/pulls?q=is%3Apr+reviewed-by%3ADiLescure" title="Reviewed Pull Requests">👀</a> <a href="#design-DiLescure" title="Design">🎨</a></td>
  </tr>
</table>
<!-- markdownlint-enable -->
<!-- prettier-ignore-end -->
<!-- ALL-CONTRIBUTORS-LIST:END -->

## License

Copyright (c) 2020-Present [Session SSO Contributors](https://github.com/simplyhexagonal/session-sso/#contributors-).<br/>
Licensed under the [Apache License 2.0](https://www.apache.org/licenses/LICENSE-2.0).


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v2.0.3](https://github.com/simplyhexagonal/session-sso/compare/v2.0.2...v2.0.3)

- chore: build dist for release [`4c7a4bd`](https://github.com/simplyhexagonal/session-sso/commit/4c7a4bd9a84ffbf5def4f8f18ef9c8facd26c49a)
- fix: header missing kid [`20abd5e`](https://github.com/simplyhexagonal/session-sso/commit/20abd5ed38d2f983625022173c7197daa1f3f9e4)

#### [v2.0.2](https://github.com/simplyhexagonal/session-sso/compare/v2.0.1...v2.0.2)

> 1 December 2021

- Release 2.0.2 [`a10ba35`](https://github.com/simplyhexagonal/session-sso/commit/a10ba35055a772164c3065741ee2982ba167439d)
- chore: build dist for release [`bb68095`](https://github.com/simplyhexagonal/session-sso/commit/bb68095211c2a7ca4396ecdfd01506fef9f130c2)
- fix: make sure SSOUtils are accessible from js [`390e0d3`](https://github.com/simplyhexagonal/session-sso/commit/390e0d3ecda80b1dbe8cc782c1ae3d4b4375d3a7)

#### [v2.0.1](https://github.com/simplyhexagonal/session-sso/compare/v2.0.0...v2.0.1)

> 1 December 2021

- Release 2.0.1 [`941b80d`](https://github.com/simplyhexagonal/session-sso/commit/941b80db4f38291e37d30a5352edcb158f9299d5)
- chore: build dist for release [`5e4b75b`](https://github.com/simplyhexagonal/session-sso/commit/5e4b75bbc2a296654fac108e3539c72a38cb2f96)
- fix: make sure SSOUtils are accessible from js [`2ea9ff6`](https://github.com/simplyhexagonal/session-sso/commit/2ea9ff6ecde8c93a8f47a0a0acc43a33f4424846)

### [v2.0.0](https://github.com/simplyhexagonal/session-sso/compare/v1.1.5...v2.0.0)

> 19 November 2021

- chore: complete refactor to simplyhexagonal standard [`b666c1e`](https://github.com/simplyhexagonal/session-sso/commit/b666c1e16bd2c77d3f8e0167bf8a914707d92abc)
- docs: update README.md [`c968782`](https://github.com/simplyhexagonal/session-sso/commit/c968782d694e506c14affa1a61441daf7fba26cd)
- Release 2.0.0 [`1a44674`](https://github.com/simplyhexagonal/session-sso/commit/1a44674fdce581a825592f8f3d89c995e8c687d8)

#### [v1.1.5](https://github.com/simplyhexagonal/session-sso/compare/v1.1.4...v1.1.5)

> 7 May 2021

- chore: update short-unique-id [`010b777`](https://github.com/simplyhexagonal/session-sso/commit/010b77720821ebd5bba7a15d43b9687873bc1499)
- chore(release): 1.1.5 [`d355ad7`](https://github.com/simplyhexagonal/session-sso/commit/d355ad7f941e095dbd73286911cf4656e69ee9eb)

#### [v1.1.4](https://github.com/simplyhexagonal/session-sso/compare/v1.1.3...v1.1.4)

> 4 May 2021

- chore: update short-unique-id [`cb55693`](https://github.com/simplyhexagonal/session-sso/commit/cb556934d044ec75f3990e91181a48a937ac7a62)
- chore(release): 1.1.4 [`acf1a93`](https://github.com/simplyhexagonal/session-sso/commit/acf1a93a8be07f28ec7fb59bb2df001d7c813d93)

#### [v1.1.3](https://github.com/simplyhexagonal/session-sso/compare/v1.1.2...v1.1.3)

> 4 May 2021

- chore: update short-unique-id version [`b9f4ce7`](https://github.com/simplyhexagonal/session-sso/commit/b9f4ce763db8035b97bac9997eadbd2b6dd4d53b)
- chore(release): 1.1.3 [`51cfdde`](https://github.com/simplyhexagonal/session-sso/commit/51cfdde741c2ba1a583d5d2707e073562d0a1aa3)

#### [v1.1.2](https://github.com/simplyhexagonal/session-sso/compare/v1.1.1...v1.1.2)

> 4 May 2021

- chore: update short-unique-id version [`f15a3bd`](https://github.com/simplyhexagonal/session-sso/commit/f15a3bd2c4992e3cd8064421246e0ca5373255ab)
- chore(release): 1.1.2 [`1839f8e`](https://github.com/simplyhexagonal/session-sso/commit/1839f8e5430dffd0afc695687014ba4cf01594ee)

#### [v1.1.1](https://github.com/simplyhexagonal/session-sso/compare/v1.1.0...v1.1.1)

> 3 May 2021

- chore: update short-unique-id to latest version [`42f95eb`](https://github.com/simplyhexagonal/session-sso/commit/42f95ebd1f141a906d3a2ca0189794bc848e77ca)
- chore(release): 1.1.1 [`410d861`](https://github.com/simplyhexagonal/session-sso/commit/410d861f4e50f02e0df3e65457c3750e33f71c40)

#### [v1.1.0](https://github.com/simplyhexagonal/session-sso/compare/v1.0.2...v1.1.0)

> 15 April 2021

- fix: eslint error by upgrading eslint ts plugins [`9184fc3`](https://github.com/simplyhexagonal/session-sso/commit/9184fc3a4d2ba550237f1944d649ac41409e4db9)
- feat: separate public and private key payload interfaces [`f065bc2`](https://github.com/simplyhexagonal/session-sso/commit/f065bc2542c96394bb7b18010461e0639499495c)
- chore(release): 1.1.0 [`19fc8ae`](https://github.com/simplyhexagonal/session-sso/commit/19fc8aefe1e7db164684f04dcb6df90dd341f154)

#### [v1.0.2](https://github.com/simplyhexagonal/session-sso/compare/v1.0.1...v1.0.2)

> 16 December 2020

- chore(release): 1.0.2 [`094d222`](https://github.com/simplyhexagonal/session-sso/commit/094d2220ce82a2e774955b4e1f8e82c6c2651075)
- docs: fix link typo [`d01d1c1`](https://github.com/simplyhexagonal/session-sso/commit/d01d1c11f82cac9c25be93cabb2064a5b33c98ec)

#### [v1.0.1](https://github.com/simplyhexagonal/session-sso/compare/v1.0.0...v1.0.1)

> 16 December 2020

- docs: docs: add better meta and amend contributor info [`039acf8`](https://github.com/simplyhexagonal/session-sso/commit/039acf8a132e8b5a9b74cc014c944837ddbdaf74)
- chore(release): 1.0.1 [`580ad9d`](https://github.com/simplyhexagonal/session-sso/commit/580ad9deca16e6257faeb4819bb14021746be0ef)

### [v1.0.0](https://github.com/simplyhexagonal/session-sso/compare/v0.9.1...v1.0.0)

> 4 November 2020

- chore: add npmignore [`c04faa3`](https://github.com/simplyhexagonal/session-sso/commit/c04faa3595fbcf1174bae408bdfd5f53d7ed5e5b)
- chore(release): 1.0.0 [`5bfd420`](https://github.com/simplyhexagonal/session-sso/commit/5bfd42016a8842d88ade8761606ec17a2871fa7c)

#### v0.9.1

> 4 November 2020

- initial commit [`a0dca82`](https://github.com/simplyhexagonal/session-sso/commit/a0dca82c9088ff0d549dd14188be48b41186c3e9)

runkit.ts
import SessionSSO, { PEMKeyPromisePayload, SSOPromiseResultFail, SSOPromiseResultSuccess } from 'session-sso';
import fetch from 'node-fetch';

const sso = new SessionSSO({
  // facebook dev credentials
  appId: '',
  appSecret: '',
  // github dev credentials
  clientId: '',
  clientSecret: '',
});

(async () => {
  const authorizationPromise = Promise.resolve({
    iss: Math.floor(Date.now() / 1000),
    exp: Math.floor(Date.now() / 1000) + (60 * 60),
    email: 'user-email-address@gmail.com',
  });

  const privateKeyPromise: Promise<PEMKeyPromisePayload> = fetch(
    'https://your.static.website/certs.priv.json',
    {
      headers: {
        // pro-tip: you can use "aws:Referer" in S3 bucket policy to restrict access to files by
        // sending a token via the HTTP Referer Header ;)
        // https://docs.aws.amazon.com/AmazonS3/latest/dev/example-bucket-policies.html#example-bucket-policies-use-case-4
        'Referer': 'WXtPUQSIc4Rp1fMOjhcRnqFR',
      },
    },
  ).then((res) => res.json()).then((jpems) => {
    const randomKeyId = Object.keys(jpems).sort(() => Math.random() - 0.5)[0];

    return {
      kid: randomKeyId,
      pem: jpems[randomKeyId],
    };
  });

  const ssoGenResult = await sso.generateSSO({
    authorizationPromise,
    privateKeyPromise,
  });

  if ((ssoGenResult as SSOPromiseResultFail).error) {
    throw new Error((ssoGenResult as SSOPromiseResultFail).error);
  }

  const {
    payload: {
      token,
    },
  } = ssoGenResult as SSOPromiseResultSuccess;

  const publicKeyPromise: Promise<PEMKeyPromisePayload> = fetch(
    'https://your.static.website/certs.json',
    {
      headers: {
        // you can use "aws:Referer" in S3 bucket policy to restrict access to files ;)
        // https://docs.aws.amazon.com/AmazonS3/latest/dev/example-bucket-policies.html#example-bucket-policies-use-case-4
        'Referer': 'WXtPUQSIc4Rp1fMOjhcRnqFR',
      },
    },
  ).then((res) => res.json());

  const verifyCustomResult = await sso.verifySSO({
    publicKeyPromise,
    authKey: token as string,
  });

  // const verifyGoogleResult = await sso.verifySSO({
  //   provider: 'google',
  //   authKey: '', // token returned by `googleUser.getAuthResponse().id_token`
  // });

  // const verifyFacebookResult = await sso.verifySSO({
  //   provider: 'facebook',
  //   authKey: '', // token returned by `FB.getLoginStatus()` => `response.authResponse.accessToken`
  // });

  // const verifyGithubResult = await sso.verifySSO({
  //   provider: 'github',
  //   authKey: '', // auth code placed in url when redirected back from https://github.com/login/oauth/authorize
  // });

  console.log(verifyCustomResult);
  // console.log(verifyGoogleResult);
  // console.log(verifyFacebookResult);
  // console.log(verifyGithubResult);
})();
